### ISSUE TITLE
__attribute__((nonstring)) and -Wunterminated-string-initialization

### ISSUE THREAD

**Issue by frobtech** [llvm/llvm-project#137705]
GCC has `-Wunterminated-string-initialization` and `__attribute__((nonstring))` to counteract it.
The warning is included in `-Wextra`.

It warns for e.g. `char foo[3] = "foo";` because the string will not have a `'\0'` terminator.
It doesn't warn for `__attribute__((nonstring)) char foo[3] = "foo";`.

This would be a useful feature to add to Clang.

**Comment by shafik** [comment]
Note, this is a C-ism, in C++ this is ill-formed.

**Comment by shafik** [comment]
CC @AaronBallman

**Comment by AaronBallman** [comment]
The attribute seems to only silence the diagnostic when `-Wc++-compat` is not enabled: https://godbolt.org/z/97Pn8Ksab

**Comment by nathanchance** [comment]
It does not look like Clang's version implements the change that GCC made for multidimensional arrays (linked at the bottom), which could make it hard for the Linux kernel to silence expected instances of this ([even if Linus hates it](https://git.kernel.org/linus/9d7a0577c9db35c4cc52db90bc415ea248446472)):

https://git.kernel.org/linus/b688f369ae0d5d25865f5441fa62e54c7d5d0de6
https://lore.kernel.org/20250415232028.work.843-kees@kernel.org/

```
$ cat test.c
const char foo[][3] = { "FOO", "BAR", };
const char bar[][3] __attribute__((__nonstring__)) = { "BAR", "FOO", };

$ clang -fsyntax-only -Wextra test.c
test.c:1:25: warning: initializer-string for character array is too long, array size is 3 but initializer has size 4 (including the null terminating character); did you mean to use the 'nonstring' attribute? [-Wunterminated-string-initialization]
    1 | const char foo[][3] = { "FOO", "BAR", };
      |                         ^~~~~
test.c:1:32: warning: initializer-string for character array is too long, array size is 3 but initializer has size 4 (including the null terminating character); did you mean to use the 'nonstring' attribute? [-Wunterminated-string-initialization]
    1 | const char foo[][3] = { "FOO", "BAR", };
      |                                ^~~~~
test.c:2:56: warning: initializer-string for character array is too long, array size is 3 but initializer has size 4 (including the null terminating character); did you mean to use the 'nonstring' attribute? [-Wunterminated-string-initialization]
    2 | const char bar[][3] __attribute__((__nonstring__)) = { "BAR", "FOO", };
      |                                                        ^~~~~
test.c:2:63: warning: initializer-string for character array is too long, array size is 3 but initializer has size 4 (including the null terminating character); did you mean to use the 'nonstring' attribute? [-Wunterminated-string-initialization]
    2 | const char bar[][3] __attribute__((__nonstring__)) = { "BAR", "FOO", };
      |                                                               ^~~~~
4 warnings generated.

$ gcc -fsyntax-only -Wextra test.c
test.c:1:25: warning: initializer-string for array of ‘char’ truncates NUL terminator but destination lacks ‘nonstring’ attribute (4 chars into 3 available) [-Wunterminated-string-initialization]
    1 | const char foo[][3] = { "FOO", "BAR", };
      |                         ^~~~~
test.c:1:32: warning: initializer-string for array of ‘char’ truncates NUL terminator but destination lacks ‘nonstring’ attribute (4 chars into 3 available) [-Wunterminated-string-initialization]
    1 | const char foo[][3] = { "FOO", "BAR", };
      |                                ^~~~~
```

https://gcc.gnu.org/g:ab714e60870d1bb98039652b760918b5e680ac10
https://gcc.gnu.org/g:1301e18f69ced6a14a7648a24d1eb3addd836b6c
https://gcc.gnu.org/g:afb46540d3921e96c4cd7ba8fa2c8b0901759455
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=117178

**Comment by AaronBallman** [comment]
Thank you for letting me know! I've posted https://github.com/llvm/llvm-project/pull/138133 which should help there, can you try it out and let me know if it addresses your concerns?
